src/rdkafka.h

```diff
@@ -6240,6 +6246,351 @@ rd_kafka_oauthbearer_set_token_failure (rd_kafka_t *rk, const char *errstr);
+ *
+ * @returns RD_KAFKA_RESP_ERR_NO_ERROR on success,
+ *          RD_KAFKA_RESP_ERR__TIMED_OUT if the transaction coordinator
+ *          could be not be contacted within \p timeout_ms (retryable),
+ *          RD_KAFKA_RESP_ERR_COORDINATOR_NOT_AVAILABLE if the transaction
+ *          coordinator is not available (retryable),
+ *          RD_KAFKA_RESP_ERR_CONCURRENT_TRANSACTIONS if a previous transaction
+ *          would not complete within \p timeout_ms (retryable),
+ *          RD_KAFKA_RESP_ERR__STATE if transactions have already been started
+ *          or upon fatal error,
+ *          RD_KAFKA_RESP_ERR__NOT_CONFIGURED if transactions have not been
+ *          configured for the producer instance,
+ *          RD_KAFKA_RESP_ERR__INVALID_ARG if \p rk is not a producer instance,
+ *          or \p timeout_ms is out of range.
+ */
+RD_EXPORT
+rd_kafka_resp_err_t
+rd_kafka_init_transactions (rd_kafka_t *rk, int timeout_ms,
+                            char *errstr, size_t errstr_size);
+
+
+
+/**
+ * @brief Begin a new transaction.
+ *
+ * rd_kafka_init_transactions() must have been called successfully (once)
+ * before this function is called.
+ *
+ * Any messages produced, offsets sent (rd_kafka_send_offsets_to_transaction()),
+ * etc, after the successful return of this function will be part of
+ * the transaction and committed or aborted atomatically.
+ *
+ * Finish the transaction by calling rd_kafka_commit_transaction() or
+ * abort the transaction by calling rd_kafka_abort_transaction().
+ *
+ * @param rk Producer instance.
+ * @param errstr A human readable error string (nul-terminated) is written to
+ *               this location that must be of at least \p errstr_size bytes.
+ *               The \p errstr is only written to if there is a fatal error.
+ * @param errstr_size Writable size in \p errstr.
+ *
+ * @returns RD_KAFKA_RESP_ERR_NO_ERROR on success,
+ *          RD_KAFKA_RESP_ERR__STATE if a transaction is already in progress
+ *          or upon fatal error,
+ *          RD_KAFKA_RESP_ERR__NOT_CONFIGURED if transactions have not been
+ *          configured for the producer instance,
+ *          RD_KAFKA_RESP_ERR__INVALID_ARG if \p rk is not a producer instance.
+ *
+ */
+RD_EXPORT
+rd_kafka_resp_err_t rd_kafka_begin_transaction (rd_kafka_t *rk,
+                                                char *errstr,
+                                                size_t errstr_size);
+
+
+/**
+ * @brief Sends a list of topic partition offsets to the consumer group
+ *        coordinator for \p consumer_group_id, and marks the offsets as part
+ *        part of the current transaction.
+ *        These offsets will be considered committed only if the transaction is
+ *        committed successfully.
+ *
+ *        The offsets should be the next message your application will consume,
+ *        i.e., the last processed message's offset + 1 for each partition.
```

Comment:
this is really dangerous - people are going to screw this up by default because everyone expects committed offset to be the last messages successfully read (and it's what you naturally have in your code without doing any arithmetic, which people won't do by default unless they're thinking hard about it, which they won't be). when offsets are committed using the consumer, this isn't such a big deal because the consumer does the calculation for you behind the scenes (for most variants). But in this case, the user will be in the loop.

given this, i think in the higher level bindings we should have an overload of `SendOffsetsToTransaction` which accepts a consumer instance. the java producer doesn't have this.

Reply:
I hear you, but we have existing offset interfaces that take the actual offset (commit, store_offsets), without doing +1, so we need to careful with being consistent in this regard.

Best thing would be if users did not have to care about individual partitions and offsets, but instead, as you propose, provide a helper method taking a consumer.

In fact, we really need three things:
 1. set start-of-transaction snapshot/restore point
 2. commit/send-offsets for transaction
 3. rewind to snapshot/restore point on transaction abort


Which would allow you to do something along the lines of:
```python

while the_truth:
    producer.begin_transaction()
    consumer.set_restorepoint()

  try:
    while time_in_txn < max_time_per_txn:
       input = consumer.poll(remaining_time)

       if consumer.is_rebalancing():
           producer.abort_transaction()
           consumer.clear_restorepoint()
            continue outer_loop

       output = process(input)
       producer.produce(output)

    producer.commit_offsets(consumer)
    producer.commit_transaction()

  except AbortableError:
          producer.abort_transaction()
          consumer.rewind_to_restorepoint()
          continue outer_loop

```

Reply:
It would be nice to expose `rd_kafka_is_transactional` for such cases so we can fail fast on eos consumer instantiation.

Reply:
It's not terribly clear to me from the mega design doc what happens in the event of a rebalance when a transaction is still open. Do we need to abort as implied by the snipped above, or is there a way for a consumer to pick up the last offset sent to a transaction for a given partition upon assignment.

Reply:
Mystery solved 

https://cwiki.apache.org/confluence/display/KAFKA/KIP-447%3A+Producer+scalability+for+exactly+once+semantics

Reply:
> It would be nice to expose rd_kafka_is_transactional for such cases so we can fail fast on eos consumer instantiation.

Can you elaborate on this? Where and how would it be used?